src/shims/unix/unnamed_socket.rs

```diff
@@ -100,33 +106,21 @@ impl FileDescription for AnonSocket {
             // corresponding ErrorKind variant.
             throw_unsup_format!("reading from the write end of a pipe");
         };
-        if readbuf.borrow().buf.is_empty() {
-            if self.peer_fd().upgrade().is_none() {
-                // Socketpair with no peer and empty buffer.
-                // 0 bytes successfully read indicates end-of-file.
-                return ecx.return_read_success(ptr, &[], 0, dest);
-            } else {
-                if self.is_nonblock {
-                    // Non-blocking socketpair with writer and empty buffer.
-                    // https://linux.die.net/man/2/read
-                    // EAGAIN or EWOULDBLOCK can be returned for socket,
-                    // POSIX.1-2001 allows either error to be returned for this case.
-                    // Since there is no ErrorKind for EAGAIN, WouldBlock is used.
-                    return ecx.set_last_error_and_return(ErrorKind::WouldBlock, dest);
-                } else {
-                    // Blocking socketpair with writer and empty buffer.
-                    // FIXME: blocking is currently not supported
-                    throw_unsup_format!("socketpair/pipe/pipe2 read: blocking isn't supported yet");
-                }
-            }
+
+        if readbuf.borrow().buf.is_empty() && self.is_nonblock {
```

Comment:
The separation of concerns between read and anonsocket_read is a bit unclear. Why is this one case handled here and everything else handled below?

I think the code would be easier to follow if you moved this logic into anonsocket_read as well.

Reply:
My idea is keeping the ``non-blocking && error`` logic in ``read`` and ``write``. In ``anonsocket_read`` and ``anonsocket_write``, we either go through the ``block_thread`` path or the actual ``read/write`` path. 

For now, ``if readbuf.borrow().buf.is_empty() && self.is_nonblock`` check will never be hit by blocking fd (the one without ``SOCK_NONBLOCK flag``) after unblocked, so I just kept them in ``read/write``. 

It is fine totally fine to move it in ``anonsocket_read`` too if it is clearer.

Reply:
Splitting the logic across two functions leads to a very odd precondition for `anonsocket_read`. It also means I have to scroll up and down to look at the entire logic. The logic really isn't that complicated so IMO it'd be better to have it in a single function where one can easily see the entire thing at a single glance.

Reply:
This makes sense, I can combine them together.

Reply:
Actually the new logic here is just wrong. When the buffer is empty, we have to *first* check `self_anonsocket.peer_fd().upgrade()` and then check ``is_nonblock``.

In the future please just keep the logic the way it is rather than trying to break it apart into two functions. :)